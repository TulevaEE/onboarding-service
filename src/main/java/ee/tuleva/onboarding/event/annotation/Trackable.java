package ee.tuleva.onboarding.event.annotation;

import static java.lang.annotation.ElementType.METHOD;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

import ee.tuleva.onboarding.event.TrackableEventType;
import java.lang.annotation.Retention;
import java.lang.annotation.Target;

@Target(METHOD)
@Retention(RUNTIME)
public @interface Trackable {
  TrackableEventType value();
}
