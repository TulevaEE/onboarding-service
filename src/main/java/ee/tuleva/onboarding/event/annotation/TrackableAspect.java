package ee.tuleva.onboarding.event.annotation;

import static java.util.stream.Collectors.toMap;

import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.ObjectMapper;
import ee.tuleva.onboarding.auth.principal.Person;
import ee.tuleva.onboarding.event.TrackableEventPublisher;
import java.util.Map;
import java.util.stream.IntStream;
import lombok.RequiredArgsConstructor;
import org.aspectj.lang.JoinPoint;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Before;
import org.aspectj.lang.reflect.MethodSignature;
import org.springframework.stereotype.Component;

@Aspect
@Component
@RequiredArgsConstructor
public class TrackableAspect {

  private final TrackableEventPublisher trackableEventPublisher;

  private final ObjectMapper objectMapper;

  @Before("@annotation(trackable) && args(.., person)")
  public void track(JoinPoint joinPoint, Trackable trackable, Person person) {
    String[] argNames = ((MethodSignature) joinPoint.getSignature()).getParameterNames();
    Object[] args = joinPoint.getArgs();

    Map<String, Object> params =
        IntStream.range(0, args.length - 1).boxed().collect(toMap(i -> argNames[i], i -> args[i]));

    JavaType type =
        objectMapper.getTypeFactory().constructMapType(Map.class, String.class, Object.class);
    Map<String, Object> data = objectMapper.convertValue(params, type);
    trackableEventPublisher.publish(person, trackable.value(), data);
  }
}
