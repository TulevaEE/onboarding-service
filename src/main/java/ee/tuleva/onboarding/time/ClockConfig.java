package ee.tuleva.onboarding.time;

import java.time.Clock;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class ClockConfig {

  @Bean
  public Clock clock() {
    return ClockHolder.clock();
  }
}
