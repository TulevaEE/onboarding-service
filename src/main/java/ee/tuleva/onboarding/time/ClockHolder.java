package ee.tuleva.onboarding.time;

import java.time.Clock;
import java.time.ZoneId;

public class ClockHolder {
  public static Clock CLOCK() {
    ZoneId timeZone = ZoneId.systemDefault();
    return Clock.system(timeZone);
  }
}
