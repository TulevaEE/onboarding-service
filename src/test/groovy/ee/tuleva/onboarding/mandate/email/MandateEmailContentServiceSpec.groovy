package ee.tuleva.onboarding.mandate.email

import au.com.origin.snapshots.Expect
import au.com.origin.snapshots.annotations.SnapshotName
import au.com.origin.snapshots.spock.EnableSnapshots
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import spock.lang.Specification

import java.time.Instant

import static ee.tuleva.onboarding.auth.UserFixture.sampleUser
import static ee.tuleva.onboarding.auth.UserFixture.sampleUserNonMember
import static ee.tuleva.onboarding.mandate.MandateFixture.sampleMandate

@SpringBootTest
@EnableSnapshots
class MandateEmailContentServiceSpec extends Specification {

  @Autowired
  MandateEmailContentService emailContentService

  private Expect expect

  @SnapshotName("second_pillar_suggest_third")
  def "#language: second pillar message: suggest third pillar when not fully converted to tuleva and not a member (third pillar active)"() {
    given:
    def user = sampleUserNonMember().build()
    def pillarSuggestion = Mock(PillarSuggestion)
    Instant mandateDate = Instant.parse("2021-11-30T10:00:00Z")
    pillarSuggestion.isSuggestPillar() >> true
    pillarSuggestion.isSuggestMembership() >> false

    when:
    String html = emailContentService.getSecondPillarHtml(
        user, mandateDate, pillarSuggestion, Locale.forLanguageTag(language))

    then:
    expect.scenario(language).toMatchSnapshot(html)

    where:
    language | _
    'en'     | _
    'et'     | _
  }

  @SnapshotName("second_pillar_suggest_membership")
  def "#language: second pillar message: suggest membership when fully converted to tuleva and not a member"() {
    given:
    def user = sampleUserNonMember().build()
    def pillarSuggestion = Mock(PillarSuggestion)
    Instant mandateDate = Instant.parse("2021-11-30T10:00:00Z")
    pillarSuggestion.isSuggestPillar() >> false
    pillarSuggestion.isSuggestMembership() >> true

    when:
    String html = emailContentService.getSecondPillarHtml(user, mandateDate, pillarSuggestion, Locale.forLanguageTag(language))

    then:
    expect.scenario(language).toMatchSnapshot(html)

    where:
    language | _
    'en'     | _
    'et'     | _
  }

  @SnapshotName("second_pillar_transfer_cancellation")
  def "#language: renders 2nd pillar transfer cancellation email"() {
    given:
    def user = sampleUserNonMember().build()
    def mandate = sampleMandate()

    when:
    String html = emailContentService.getSecondPillarTransferCancellationHtml(user, mandate, Locale.forLanguageTag(language))

    then:
    expect.scenario(language).toMatchSnapshot(html)

    where:
    language | _
    'en'     | _
    'et'     | _
  }

  @SnapshotName("second_pillar_withdraw_cancellation")
  def "#language: renders 2nd pillar withdrawal cancellation email"() {
    given:
    def user = sampleUserNonMember().build()
    def mandate = sampleMandate()

    when:
    String html = emailContentService.getSecondPillarWithdrawalCancellationHtml(user, Locale.forLanguageTag(language))

    then:
    expect.scenario(language).toMatchSnapshot(html)

    where:
    language | _
    'en'     | _
    'et'     | _
  }

  @SnapshotName("third_pillar_payment_reminder")
  def "#language: renders third pillar mandate payment reminder email correctly"() {
    given:
    def user = sampleUser().build()

    when:
    String html = emailContentService.getThirdPillarPaymentReminderHtml(user, Locale.forLanguageTag(language))

    then:
    expect.scenario(language).toMatchSnapshot(html)

    where:
    language | _
    'en'     | _
    'et'     | _
  }
}
