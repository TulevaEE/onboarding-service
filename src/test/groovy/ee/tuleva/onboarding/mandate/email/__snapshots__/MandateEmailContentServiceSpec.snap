second_pillar_suggest_membership[en]=[
<!DOCTYPE HTML>
<html>
  <body>
    <p>Hello Jordan,</p>
    <br>
    <p>You are now saving for your pension alongside me and other Tuleva members. Your payments will be directed to the chosen pension fund from the next payment made. If you wish to swap any existing shares, the exchange of units will take effect on 03.01.2022.</p>
    <br>
    
    <p>Tuleva's owners are the pension savers themselves. As a Tuleva pension fund client, you have no obligation to become a member of Tuleva. All Estonian citizens are able to save together with us in our modern, low-cost pension funds. Having said that, I would still welcome you to think about becoming a member, since you benefit the most from Tuleva whilst being a co-owner yourself.<br><br><a href="https://tuleva.ee/en/mutual-company-2/?utm_source=autoemail&utm_medium=clients&utm_campaign=sug_mem">More information about the benefits of being a member of Tuleva</a></p>
    <br>
    <p>Attached is a copy of the application you just made on our website and which we have sent to Pensionikeskus.</p>
    <p>If you have any questions please contact us via <a href="mailto:tuleva@tuleva.ee">tuleva@tuleva.ee</a> or 644 5100.</p>
    <br>
    <p>Tuleva</p>
  </body>
</html>

]


second_pillar_suggest_membership[et]=[
<!DOCTYPE HTML>
<html>
  <body>
    <p>Tervist, Jordan.</p>
    <br>
    <p>Kogud nüüd oma teise samba pensionit koos minu ja teiste Tuleva liikmetega. Sinu maksed suunatakse valitud pensionifondi alates järgmisest maksest. Kui sa soovisid olemasolevad osakud vahetada, siis osakute vahetus jõustub 03.01.2022.</p>
    <br>
    
    <p>Tuleva pensionifondide valitseja kuulub inimestele, kes ka ise neis fondides oma tuleviku jaoks raha koguvad. Tuleva pensionifondide kliendina ei ole sul mingit kohustust Tuleva ühistu liikmeks astuda - meiega koos võivad kaasaegses, madalate kuludega fondis pensionit koguda kõik Eesti inimesed. Aga kutsun sind liikmeks tulemist kaaluma, sest kõige rohkem saad Tulevast kasu, olles ise kaasomanik.<br><br><a href="https://tuleva.ee/liikmetele/31-marts-korduma-kippuvad-kusimused/?utm_source=autoemail&utm_medium=clients&utm_campaign=sug_mem">Vaata lähemalt, mida Tuleva liikmeks olek tähendab</a></p>
    <br>
    <p>Käesolevale kirjale on lisatud koopia sinu tehtud avaldustest, mis on saadetud Pensionikeskusesse.</p>
    <p>Kui sul tekib ükskõik milliseid küsimusi, võta kõhklemata ühendust: e-post <a href="mailto:tuleva@tuleva.ee">tuleva@tuleva.ee</a>, tel 644 5100.</p>
    <br>
    <p>Tuleva</p>
  </body>
</html>

]


second_pillar_suggest_third[en]=[
<!DOCTYPE HTML>
<html>
  <body>
    <p>Hello Jordan,</p>
    <br>
    <p>You are now saving for your pension alongside me and other Tuleva members. Your payments will be directed to the chosen pension fund from the next payment made. If you wish to swap any existing shares, the exchange of units will take effect on 03.01.2022.</p>
    <br>
    <p>Next, set up your third pillar. You receive back the income tax on any contributions made this year already next spring. <a href="https://tuleva.ee/iii-sammas/?utm_source=autoemail&utm_medium=clients&utm_campaign=sug_3">Start from here.</a></p>
    
    <br>
    <p>Attached is a copy of the application you just made on our website and which we have sent to Pensionikeskus.</p>
    <p>If you have any questions please contact us via <a href="mailto:tuleva@tuleva.ee">tuleva@tuleva.ee</a> or 644 5100.</p>
    <br>
    <p>Tuleva</p>
  </body>
</html>

]


second_pillar_suggest_third[et]=[
<!DOCTYPE HTML>
<html>
  <body>
    <p>Tervist, Jordan.</p>
    <br>
    <p>Kogud nüüd oma teise samba pensionit koos minu ja teiste Tuleva liikmetega. Sinu maksed suunatakse valitud pensionifondi alates järgmisest maksest. Kui sa soovisid olemasolevad osakud vahetada, siis osakute vahetus jõustub 03.01.2022.</p>
    <br>
    <p>Järgmisena sea üles kolmas sammas. Sel aastal tehtud sissemaksetelt annab riik sulle tulumaksu juba kevadel tagasi. <a href="https://tuleva.ee/iii-sammas/?utm_source=autoemail&utm_medium=clients&utm_campaign=sug_3">Alusta siit.</a></p>
    
    <br>
    <p>Käesolevale kirjale on lisatud koopia sinu tehtud avaldustest, mis on saadetud Pensionikeskusesse.</p>
    <p>Kui sul tekib ükskõik milliseid küsimusi, võta kõhklemata ühendust: e-post <a href="mailto:tuleva@tuleva.ee">tuleva@tuleva.ee</a>, tel 644 5100.</p>
    <br>
    <p>Tuleva</p>
  </body>
</html>

]


second_pillar_transfer_cancellation[en]=[
<!DOCTYPE HTML>
<html>
  <body>
    <p>Hello Jordan,</p>
    <br>
    <p>You have submitted a cancellation application through Tuleva. Your second pillar assets will stay in Kohustuslik Pensionifond Danske Pension 25.</p>
    <br>
    <p>Attached is a copy of the application you just made on our website and which we have sent to Pensionikeskus.</p>
    <p>If you have any questions please contact us via <a href="mailto:tuleva@tuleva.ee">tuleva@tuleva.ee</a> or 644 5100.</p>
    <br>
    <p>Tuleva</p>
  </body>
</html>

]


second_pillar_transfer_cancellation[et]=[
<!DOCTYPE HTML>
<html>
  <body>
    <p>Tervist, Jordan.</p>
    <br>
    <p>Tegid Tuleva kaudu tühistamise avalduse. Sinu teise samba vara jääb edasi fondi: Kohustuslik Pensionifond Danske Pension 25.</p>
    <br>
    <p>Käesolevale kirjale on lisatud koopia sinu tehtud avaldustest, mis on saadetud Pensionikeskusesse.</p>
    <p>Kui sul tekib ükskõik milliseid küsimusi, võta kõhklemata ühendust: e-post <a href="mailto:tuleva@tuleva.ee">tuleva@tuleva.ee</a>, tel 644 5100.</p>
    <br>
    <p>Tuleva</p>
  </body>
</html>

]


second_pillar_withdraw_cancellation[en]=[
<!DOCTYPE HTML>
<html>
  <body>
    <p>Hello Jordan,</p>
    <br>
    <p>You have cancelled your 2nd pillar withdrawal application. You continue to save in 2nd pillar as before.</p>
    <br>
    <p>Attached is a copy of the application you just made on our website and which we have sent to Pensionikeskus.</p>
    <p>If you have any questions please contact us via <a href="mailto:tuleva@tuleva.ee">tuleva@tuleva.ee</a> or 644 5100.</p>
    <br>
    <p>Tuleva</p>
  </body>
</html>

]


second_pillar_withdraw_cancellation[et]=[
<!DOCTYPE HTML>
<html>
  <body>
    <p>Tervist, Jordan.</p>
    <br>
    <p>Tühistasid Tuleva kaudu oma teisest sambast lahkumise avalduse ja sa jätkad kogumist teises sambas nii nagu seni.</p>
    <br>
    <p>Käesolevale kirjale on lisatud koopia sinu tehtud avaldustest, mis on saadetud Pensionikeskusesse.</p>
    <p>Kui sul tekib ükskõik milliseid küsimusi, võta kõhklemata ühendust: e-post <a href="mailto:tuleva@tuleva.ee">tuleva@tuleva.ee</a>, tel 644 5100.</p>
    <br>
    <p>Tuleva</p>
  </body>
</html>

]


third_pillar_payment_reminder[en]=[
<!DOCTYPE HTML>
<html>
  <body>
    <p>Welcome Jordan,</p>
    <p>I’m glad you’ve decided to start saving in the third pillar with us.</p>
    <p>Your contribution has not reached the Tuleva III Pillar Pension Fund yet. If you made a payment from your internet bank or set up a recurring payment, your contribution will show up on your pension account one working day after payment.</p>
    <p>If you forgot to make a payment, <a href="https://pension.tuleva.ee/3rd-pillar-flow/payment?utm_source=autoemail&utm_medium=clients&utm_campaign=reminder_3">log into your pension account</a> and do it now.</p>
    <p>Attached is a copy of your third pillar application, which we have sent to the Pension Registry.</p>

    <p>If you have any questions please contact us via <a href="mailto:tuleva@tuleva.ee">tuleva@tuleva.ee</a> or 644 5100.</p>
    <br/>
    <p>Tõnu Pekk<br>Fund manager<br>Tuleva</p>
  </body>
</html>

]


third_pillar_payment_reminder[et]=[
<!DOCTYPE HTML>
<html>
  <body>
    <p>Tere tulemast, Jordan.</p>
    <p>Mul on hea meel, et oled alustanud meiega koos kolmandasse sambasse raha kogumist.</p>
    <p>Sinu sissemakse pole veel Tuleva III Samba Pensionifondi jõudnud. Kui tegid makse oma internetipangast või seadsid üles püsimakse, siis jõuab sinu raha fondi üks tööpäev pärast makse tegemist. Siis näed makset ka oma pensionikontol.</p>
    <p>Kui unustasid makse tegemata, <a href="https://pension.tuleva.ee/3rd-pillar-flow/payment?utm_source=autoemail&utm_medium=clients&utm_campaign=reminder_3">logi sisse oma pensionikontole</a> ja tee makse ära.</p>
    <p>Manuses on koopia sinu tehtud avaldustest, mille edastasime Pensionikeskusele.</p>

    <p>Kui sul tekib ükskõik milliseid küsimusi, võta kõhklemata ühendust: e-post <a href="mailto:tuleva@tuleva.ee">tuleva@tuleva.ee</a>, tel 644 5100.</p>
    <br/>
    <p>Tõnu Pekk<br>Fondijuht<br>Tuleva</p>
  </body>
</html>

]
